ledger/accountdb.go

```diff
@@ -633,6 +644,406 @@ func accountsAddNormalizedBalance(tx *sql.Tx, proto config.ConsensusParams) erro
 	return rows.Err()
 }
 
+// accountsCreateResourceTable creates the resource table in the database.
+func accountsCreateResourceTable(ctx context.Context, tx *sql.Tx) error {
+	var exists bool
+	err := tx.QueryRowContext(ctx, "SELECT 1 FROM pragma_table_info('resources') WHERE name='addrid'").Scan(&exists)
+	if err == nil {
+		// Already exists.
+		return nil
+	}
+	if err != sql.ErrNoRows {
+		return err
+	}
+	for _, stmt := range createResourcesTable {
+		_, err = tx.ExecContext(ctx, stmt)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type baseOnlineAccountData struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	VoteID          crypto.OneTimeSignatureVerifier `codec:"A"`
+	SelectionID     crypto.VRFVerifier              `codec:"B"`
+	VoteFirstValid  basics.Round                    `codec:"C"`
+	VoteLastValid   basics.Round                    `codec:"D"`
+	VoteKeyDilution uint64                          `codec:"E"`
+}
+
+type baseAccountData struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Status                     basics.Status     `codec:"a"`
+	MicroAlgos                 basics.MicroAlgos `codec:"b"`
+	RewardsBase                uint64            `codec:"c"`
+	RewardedMicroAlgos         basics.MicroAlgos `codec:"d"`
+	AuthAddr                   basics.Address    `codec:"e"`
+	TotalAppSchemaNumUint      uint64            `codec:"f"`
+	TotalAppSchemaNumByteSlice uint64            `codec:"g"`
+	TotalExtraAppPages         uint32            `codec:"h"`
+
+	baseOnlineAccountData
+
+	UpdateRound uint64 `codec:"z"`
+}
+
+func (ba *baseAccountData) SetAccountData(ad *basics.AccountData) {
+	ba.Status = ad.Status
+	ba.MicroAlgos = ad.MicroAlgos
+	ba.RewardsBase = ad.RewardsBase
+	ba.RewardedMicroAlgos = ad.RewardedMicroAlgos
+	ba.VoteID = ad.VoteID
+	ba.SelectionID = ad.SelectionID
+	ba.VoteFirstValid = ad.VoteFirstValid
+	ba.VoteLastValid = ad.VoteLastValid
+	ba.VoteKeyDilution = ad.VoteKeyDilution
+	ba.AuthAddr = ad.AuthAddr
+	ba.TotalAppSchemaNumUint = ad.TotalAppSchema.NumUint
+	ba.TotalAppSchemaNumByteSlice = ad.TotalAppSchema.NumByteSlice
+	ba.TotalExtraAppPages = ad.TotalExtraAppPages
+}
+
+func (ba *baseAccountData) GetAccountData() basics.AccountData {
+	return basics.AccountData{
+		Status:             ba.Status,
+		MicroAlgos:         ba.MicroAlgos,
+		RewardsBase:        ba.RewardsBase,
+		RewardedMicroAlgos: ba.RewardedMicroAlgos,
+		VoteID:             ba.VoteID,
+		SelectionID:        ba.SelectionID,
+		VoteFirstValid:     ba.VoteFirstValid,
+		VoteLastValid:      ba.VoteLastValid,
+		VoteKeyDilution:    ba.VoteKeyDilution,
+		AuthAddr:           ba.AuthAddr,
+		TotalAppSchema: basics.StateSchema{
+			NumUint:      ba.TotalAppSchemaNumUint,
+			NumByteSlice: ba.TotalAppSchemaNumByteSlice,
+		},
+		TotalExtraAppPages: ba.TotalExtraAppPages,
+	}
+}
+
+type resourceFlags uint8
+
+const (
+	resourceFlagsHolding    = 0 //nolint:deadcode
```

Comment:
is it some kind of optimization like most of the entries would have holdings?

Reply:
exactly. Most of the entries would be for "holding". very few will be for "owning".

Reply:
Maybe simpler (and more extensible if we want to cut the resources storage differently some day) to have four flags? 0x1 AssetHoldingExists, 0x2 AssetParamsExists, 0x4 AppParamsExist, 0x8 AppLocalStateExists? It seems a little bug-prone to have implicit meaning to empty/zero values and not distinguish between different app vs asset use cases with names like "holding" and "ownership" that apply to both. I guess the argument is you can combine the with rtype to deduce the real flag, and it saves a few bytes per record to default to holding vs owning.

Reply:
I guess my main discomfort is, the zero value of this struct implies the existence of a zero-valued basics.AssetHolding{} or zero-valued basics.AppLocalState (for UpdateRound zero). Maybe that's OK as long as handled safety and hidden down here in the DB.

Reply:
Yes, my intent is to wrap it nicely so that it won't be externally exposed. But before doing that, I would need to know what the "exported" data structures are going to be ( i.e. it's not going to be AccountData ). 
I will defer this to a future PR, as this unblocks me from moving forward.